presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java

```diff
@@ -577,6 +551,152 @@ public Signature resolveFunction(QualifiedName name, List<TypeSignature> paramet
+            if (returnTypeIsTheSame(mostSpecificFunctions) && unknownParametersAreNotNullable(mostSpecificFunctions, parameters)) {
+                // make it deterministic
+                ApplicableFunction selectedFunction = Ordering.usingToString()
+                        .reverse()
+                        .sortedCopy(mostSpecificFunctions)
+                        .get(0);
+                return ImmutableList.of(selectedFunction);
+            }
+        }
+        return mostSpecificFunctions;
+    }
+
+    private List<ApplicableFunction> filterOutLessSpecificFunctions(List<ApplicableFunction> applicableFunctions)
+    {
+        List<ApplicableFunction> mostSpecificFunctions = new ArrayList<>();
+        for (ApplicableFunction function : applicableFunctions) {
+            Optional<Integer> mostSpecificId = function.getIdOfFunctionFamily(mostSpecificFunctions);
+            if (mostSpecificId.isPresent() && function.isMoreSpecificThan(mostSpecificFunctions.get(mostSpecificId.get()))) {
+                mostSpecificFunctions.set(mostSpecificId.get(), function);
+            }
+
+            if (!mostSpecificId.isPresent()) {
+                // the 'function' does not fit to any specific family
+                mostSpecificFunctions.add(function);
+            }
+        }
+        return ImmutableList.copyOf(mostSpecificFunctions);
+    }
+
+    private boolean someParameterIsUnknown(List<Type> parameters)
+    {
+        return parameters.stream().anyMatch(type -> type.equals(UNKNOWN));
+    }
+
+    private List<ApplicableFunction> getUnknownOnlyCastFunctions(List<ApplicableFunction> applicableFunction, List<Type> actualParameters)
+    {
+        return applicableFunction.stream()
+                .filter((function) -> onlyCastsUnknown(function, actualParameters))
+                .collect(toImmutableList());
+    }
+
+    private boolean onlyCastsUnknown(ApplicableFunction applicableFunction, List<Type> actualParameters)
+    {
+        List<Type> boundTypes = resolveTypes(applicableFunction.getBoundSignature().getArgumentTypes(), typeManager);
+        checkState(actualParameters.size() == boundTypes.size(), "type lists are of different lengths");
+        for (int i = 0; i < actualParameters.size(); i++) {
+            if (!boundTypes.get(i).equals(actualParameters.get(i)) && actualParameters.get(i) != UNKNOWN) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private boolean returnTypeIsTheSame(List<ApplicableFunction> applicableFunctions)
+    {
+        Set<Type> returnTypes = applicableFunctions.stream()
+                .map(function -> typeManager.getType(function.getBoundSignature().getReturnType()))
+                .collect(Collectors.toSet());
+        return returnTypes.size() == 1;
+    }
+
+    private boolean unknownParametersAreNotNullable(List<ApplicableFunction> applicableFunctions, List<Type> parameters)
```

Comment:
I'd name this "returnsNullOnNullInput", as that is really what we're trying to determine. We can only guarantee that property for functions that don't process null input on their own, which is signaled by the absence of nullable arguments.